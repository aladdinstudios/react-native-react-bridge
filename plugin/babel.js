const { parseSync, transformSync } = require("@babel/core");
const traverse = require("@babel/traverse").default;
const p = require("path");

export const extractEntryPaths = (src, filename) => {
  const ast = parseSync(src, { filename });
  const compPaths = [];
  traverse(ast, {
    CallExpression(path) {
      if (isBundleReact(path.node)) {
        const [source, ...args] = path.get("arguments");
        compPaths.push(source.node.value);
      }
    },
  });
  return compPaths;
};

export const replaceEntryPaths = (src, filename, compStrs) => {
  return transformSync(src, {
    filename,
    // babelrc: false,
    plugins: [
      {
        visitor: {
          CallExpression(path, { file: { opts: fileOpts } }) {
            if (isBundleReact(path.node)) {
              const [source, ...args] = path.get("arguments");
              path.replaceWithSourceString(compStrs[source.node.value]);
            }
          },
        },
      },
    ],
  });
};

function isBundleReact(node) {
  return (
    looksLike(node, {
      callee: {
        type: "Identifier",
        name: "bundleReact",
      },
    }) ||
    looksLike(node, {
      callee: {
        type: "MemberExpression",
        property: { name: "bundleReact" },
      },
    })
  );
}

function looksLike(a, b) {
  return (
    a &&
    b &&
    Object.keys(b).every((bKey) => {
      const bVal = b[bKey];
      const aVal = a[bKey];
      if (typeof bVal === "function") {
        return bVal(aVal);
      }
      return isPrimitive(bVal) ? bVal === aVal : looksLike(aVal, bVal);
    })
  );
}

function isPrimitive(val) {
  return val == null || /^[sbn]/.test(typeof val);
}
