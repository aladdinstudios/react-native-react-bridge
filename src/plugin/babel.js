import { parseSync, transformSync } from "@babel/core";
import traverse from "@babel/traverse";

export const isEntryFile = (src, filename) => {
  const ast = parseSync(src, { filename });
  let isEntry = false;
  traverse(ast, {
    ExportDefaultDeclaration(path) {
      if (
        looksLike(path.node, {
          declaration: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "webViewRender",
            },
          },
        })
      ) {
        isEntry = true;
      }
    },
  });
  return isEntry;
};

function looksLike(a, b) {
  return (
    a &&
    b &&
    Object.keys(b).every((bKey) => {
      const bVal = b[bKey];
      const aVal = a[bKey];
      if (typeof bVal === "function") {
        return bVal(aVal);
      }
      return isPrimitive(bVal) ? bVal === aVal : looksLike(aVal, bVal);
    })
  );
}

function isPrimitive(val) {
  return val == null || /^[sbn]/.test(typeof val);
}
