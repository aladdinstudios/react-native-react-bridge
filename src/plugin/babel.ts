import { parseSync } from "@babel/core";
import traverse from "@babel/traverse";

/**
 * @internal
 */
export const isEntryFile = (src: string, filename: string) => {
  // TODO try lighter approach
  const ast = parseSync(src, { filename })!;
  let isEntry = false;
  traverse(ast, {
    ExportDefaultDeclaration(path) {
      if (
        looksLike(path.node, {
          declaration: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "webViewRender",
            },
          },
        }) ||
        looksLike(path.node, {
          declaration: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "webViewCreateRoot",
            },
          },
        })
      ) {
        isEntry = true;
      }
    },
  });
  return isEntry;
};

function looksLike(a: any, b: any): boolean {
  return (
    a &&
    b &&
    Object.keys(b).every((bKey) => {
      const bVal = b[bKey as keyof typeof b];
      const aVal = a[bKey as keyof typeof b];
      if (typeof bVal === "function") {
        return bVal(aVal);
      }
      return isPrimitive(bVal) ? bVal === aVal : looksLike(aVal, bVal);
    })
  );
}

function isPrimitive(val: any): boolean {
  return val == null || /^[sbn]/.test(typeof val);
}
