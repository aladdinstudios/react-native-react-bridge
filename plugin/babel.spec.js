import * as fs from "fs";
import * as path from "path";
import { extractEntryPaths, replaceEntryPaths } from "./babel";

const readFixture = (filename) =>
  fs.readFileSync(path.join(__dirname, "../fixtures", filename), "utf-8");

describe("extractEntryPaths", () => {
  it("import { bundleReact }", () => {
    const filename = "app-import-destructing.jsx";
    const src = readFixture(filename);
    const paths = extractEntryPaths(src, filename);
    expect(paths).toEqual(["./Component"]);
  });

  it("import all", () => {
    const filename = "app-import-all.jsx";
    const src = readFixture(filename);
    const paths = extractEntryPaths(src, filename);
    expect(paths).toEqual(["./Component"]);
  });

  it("require { bundleReact }", () => {
    const filename = "app-require-destructing.jsx";
    const src = readFixture(filename);
    const paths = extractEntryPaths(src, filename);
    expect(paths).toEqual(["./Component"]);
  });

  it("require all", () => {
    const filename = "app-require-all.jsx";
    const src = readFixture(filename);
    const paths = extractEntryPaths(src, filename);
    expect(paths).toEqual(["./Component"]);
  });
});

describe("replaceEntryPaths", () => {
  it("import { bundleReact }", () => {
    const filename = "app-import-destructing.jsx";
    const src = readFixture(filename);
    const res = replaceEntryPaths(src, filename, {
      "./Component": "REPLACED",
    });
    expect(res.code).toMatchSnapshot();
  });

  it("import all", () => {
    const filename = "app-import-all.jsx";
    const src = readFixture(filename);
    const res = replaceEntryPaths(src, filename, {
      "./Component": "REPLACED",
    });
    expect(res.code).toMatchSnapshot();
  });

  it("require { bundleReact }", () => {
    const filename = "app-require-destructing.jsx";
    const src = readFixture(filename);
    const res = replaceEntryPaths(src, filename, {
      "./Component": "REPLACED",
    });
    expect(res.code).toMatchSnapshot();
  });

  it("require all", () => {
    const filename = "app-require-all.jsx";
    const src = readFixture(filename);
    const res = replaceEntryPaths(src, filename, {
      "./Component": "REPLACED",
    });
    expect(res.code).toMatchSnapshot();
  });
});
