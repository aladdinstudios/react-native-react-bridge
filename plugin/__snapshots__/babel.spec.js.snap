// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceEntryPaths import { bundleReact } 1`] = `"\\"use strict\\";var _interopRequireDefault=require(\\"@babel/runtime/helpers/interopRequireDefault\\");var _react=_interopRequireDefault(require(\\"react\\"));var _reactNativeReactBundler=require(\\"react-native-react-bundler\\");var Comp=REPLACED;var App=function App(){return _react.default.createElement(_reactNativeReactBundler.RNDOM,{app:Comp});};"`;

exports[`replaceEntryPaths import all 1`] = `"\\"use strict\\";var _interopRequireWildcard=require(\\"@babel/runtime/helpers/interopRequireWildcard\\");var _interopRequireDefault=require(\\"@babel/runtime/helpers/interopRequireDefault\\");var _react=_interopRequireDefault(require(\\"react\\"));var RNR=_interopRequireWildcard(require(\\"react-native-react-bundler\\"));var Comp=REPLACED;var App=function App(){return _react.default.createElement(RNR.RNDOM,{app:Comp});};"`;

exports[`replaceEntryPaths require { bundleReact } 1`] = `"\\"use strict\\";var React=require(\\"react\\");var _require=require(\\"react-native-react-bundler\\"),bundleReact=_require.bundleReact,RNDOM=_require.RNDOM;var Comp=REPLACED;var App=function App(){return React.createElement(RNDOM,{app:Comp});};"`;

exports[`replaceEntryPaths require all 1`] = `"\\"use strict\\";var React=require(\\"react\\");var RNR=require(\\"react-native-react-bundler\\");var Comp=REPLACED;var App=function App(){return React.createElement(RNR.RNDOM,{app:Comp});};"`;
